File: src/test/java/no/nav/bidrag/beregn/bidragsevne/BidragsevneberegningTest.java
package no.nav.bidrag.beregn.bidragsevne;

import static java.util.Collections.emptyList;
import static java.util.Collections.singletonList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import no.nav.bidrag.beregn.TestUtil;
import no.nav.bidrag.beregn.bidragsevne.beregning.BidragsevneberegningImpl;
import no.nav.bidrag.beregn.bidragsevne.bo.GrunnlagBeregning;
import no.nav.bidrag.beregn.bidragsevne.bo.Inntekt;
import no.nav.bidrag.beregn.felles.SjablonUtil;
import no.nav.bidrag.beregn.felles.bo.Sjablon;
import no.nav.bidrag.beregn.felles.bo.SjablonInnhold;
import no.nav.bidrag.beregn.felles.enums.BostatusKode;
import no.nav.bidrag.beregn.felles.enums.InntektType;
import no.nav.bidrag.beregn.felles.enums.SaerfradragKode;
import no.nav.bidrag.beregn.felles.enums.SjablonInnholdNavn;
import no.nav.bidrag.beregn.felles.enums.SjablonTallNavn;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

@DisplayName("BidragsevneBeregningTest")
class BidragsevneberegningTest {

  private final List<Sjablon> sjablonListe = TestUtil.byggSjabloner();

  @Test
  void beregn() {

    ArrayList<Inntekt> inntekter = new ArrayList<>();

    BidragsevneberegningImpl bidragsevneberegning = new BidragsevneberegningImpl();

    // Tester beregning med ulike inntekter
    inntekter.add(new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(1000000)));
    GrunnlagBeregning grunnlagBeregning
        = new GrunnlagBeregning(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(1),
        SaerfradragKode.INGEN, sjablonListe);
    assertEquals(BigDecimal.valueOf(31859),
        bidragsevneberegning.beregn(grunnlagBeregning).getResultatEvneBelop());

    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(520000)));
    GrunnlagBeregning grunnlagBeregning2
        = new GrunnlagBeregning(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(1),
        SaerfradragKode.INGEN, sjablonListe);
    assertEquals(BigDecimal.valueOf(8322),
        bidragsevneberegning.beregn(grunnlagBeregning2).getResultatEvneBelop());

    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(666000)));
    GrunnlagBeregning grunnlagBeregning3
        = new GrunnlagBeregning(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(3),
        SaerfradragKode.INGEN, sjablonListe);
    assertEquals(BigDecimal.valueOf(8424),
        bidragsevneberegning.beregn(grunnlagBeregning3).getResultatEvneBelop());

    // Test på at beregnet bidragsevne blir satt til 0 når evne er negativ
    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(100000)));
    GrunnlagBeregning grunnlagBeregning4
        = new GrunnlagBeregning(inntekter, 1, BostatusKode.MED_ANDRE, BigDecimal.valueOf(1),
        SaerfradragKode.HELT, sjablonListe);
    assertEquals(BigDecimal.ZERO,
        bidragsevneberegning.beregn(grunnlagBeregning4).getResultatEvneBelop());

    // Test at fordel skatteklasse 2 ikke legges til beregnet evne når skatteklasse = 1
    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(666000)));
    sjablonListe.set(0, new Sjablon(SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELOP.getNavn(), emptyList(),
        singletonList(new SjablonInnhold(SjablonInnholdNavn.SJABLON_VERDI.getNavn(), BigDecimal.valueOf(12000)))));

    GrunnlagBeregning grunnlagBeregning5
        = new GrunnlagBeregning(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(3),
        SaerfradragKode.INGEN, sjablonListe);
    assertEquals(BigDecimal.valueOf(8424),
        bidragsevneberegning.beregn(grunnlagBeregning5).getResultatEvneBelop());

    // Test at fordel skatteklasse 2 legges til beregnet evne når skatteklasse = 2
    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(666000)));
    sjablonListe.set(0, new Sjablon(SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELOP.getNavn(), emptyList(),
        singletonList(new SjablonInnhold(SjablonInnholdNavn.SJABLON_VERDI.getNavn(), BigDecimal.valueOf(12000)))));

    GrunnlagBeregning grunnlagBeregning6
        = new GrunnlagBeregning(inntekter, 2, BostatusKode.ALENE, BigDecimal.valueOf(3),
        SaerfradragKode.INGEN, sjablonListe);
    assertEquals(BigDecimal.valueOf(9424),
        bidragsevneberegning.beregn(grunnlagBeregning6).getResultatEvneBelop());

    // Test at personfradrag skatteklasse 2 brukes hvis skatteklasse 2 er angitt
    sjablonListe.set(0, new Sjablon(SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELOP.getNavn(), emptyList(),
        singletonList(new SjablonInnhold(SjablonInnholdNavn.SJABLON_VERDI.getNavn(), BigDecimal.ZERO))));

    sjablonListe.set(1, new Sjablon(SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELOP.getNavn(), emptyList(),
        singletonList(new SjablonInnhold(SjablonInnholdNavn.SJABLON_VERDI.getNavn(), BigDecimal.valueOf(24000)))));

    GrunnlagBeregning grunnlagBeregning7
        = new GrunnlagBeregning(inntekter, 2, BostatusKode.ALENE, BigDecimal.valueOf(3),
        SaerfradragKode.INGEN, sjablonListe);
    assertEquals(BigDecimal.valueOf(7923),
        bidragsevneberegning.beregn(grunnlagBeregning7).getResultatEvneBelop());


    // Test av halvt særfradrag
    GrunnlagBeregning grunnlagBeregning8
        = new GrunnlagBeregning(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(3),
        SaerfradragKode.HALVT, sjablonListe);
    assertEquals(BigDecimal.valueOf(8965),
        bidragsevneberegning.beregn(grunnlagBeregning8).getResultatEvneBelop());

    // Test av bostatus MED_FLERE
    GrunnlagBeregning grunnlagBeregning9
        = new GrunnlagBeregning(inntekter, 1, BostatusKode.MED_ANDRE, BigDecimal.valueOf(3),
        SaerfradragKode.HALVT, sjablonListe);
    assertEquals(BigDecimal.valueOf(14253),
        bidragsevneberegning.beregn(grunnlagBeregning9).getResultatEvneBelop());

  }

  @Test
  void beregnMinstefradrag() {

    ArrayList<Inntekt> inntekter = new ArrayList<>();
    inntekter.add(new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(200000)));

    BidragsevneberegningImpl bidragsevneberegning = new BidragsevneberegningImpl();

    GrunnlagBeregning grunnlagBeregning
        = new GrunnlagBeregning(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(1),
        SaerfradragKode.HELT, sjablonListe);
    System.out.println(bidragsevneberegning.beregnMinstefradrag(grunnlagBeregning,
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT)));
    assertThat(bidragsevneberegning.beregnMinstefradrag(grunnlagBeregning,
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT)))
        .isEqualTo(BigDecimal.valueOf(62000));

    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(1000000)));
    GrunnlagBeregning grunnlagBeregning2
        = new GrunnlagBeregning(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(1),
        SaerfradragKode.HELT, sjablonListe);
    System.out.println(bidragsevneberegning.beregnMinstefradrag(grunnlagBeregning2,
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT)));
    assertThat(bidragsevneberegning.beregnMinstefradrag(grunnlagBeregning2,
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT)))
        .isEqualTo(BigDecimal.valueOf(87450));
  }

  @Test
  void beregnSkattetrinnBelop() {

    BidragsevneberegningImpl bidragsevneberegning = new BidragsevneberegningImpl();

    ArrayList<Inntekt> inntekter = new ArrayList<>();
    inntekter.add(new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(666000)));

    GrunnlagBeregning grunnlagBeregning
        = new GrunnlagBeregning(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(1),
        SaerfradragKode.HELT, sjablonListe);
    assertEquals(BigDecimal.valueOf(1400+16181+3465+0),
        bidragsevneberegning.beregnSkattetrinnBelop(grunnlagBeregning));

    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(174600)));
    GrunnlagBeregning grunnlagBeregning2
        = new GrunnlagBeregning(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(1),
        SaerfradragKode.HELT, sjablonListe);
    assertEquals(BigDecimal.ZERO,
        bidragsevneberegning.beregnSkattetrinnBelop(grunnlagBeregning2));

    inntekter.set(0, new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(250000)));
    GrunnlagBeregning grunnlagBeregning3
        = new GrunnlagBeregning(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(1),
        SaerfradragKode.HELT, sjablonListe);
    assertEquals(BigDecimal.valueOf(1315),
        bidragsevneberegning.beregnSkattetrinnBelop(grunnlagBeregning3));
  }

  @Test
  void TestFraJohn() {

    ArrayList<Inntekt> inntekter = new ArrayList<>();

    BidragsevneberegningImpl bidragsevneberegning = new BidragsevneberegningImpl();

    // Tester beregning med ulike inntekter
    inntekter.add(new Inntekt(InntektType.LONN_SKE, BigDecimal.valueOf(300000)));
    GrunnlagBeregning grunnlagBeregning
        = new GrunnlagBeregning(inntekter, 1, BostatusKode.ALENE, BigDecimal.valueOf(0),
        SaerfradragKode.HELT, sjablonListe);
    assertEquals(BigDecimal.valueOf(1217),
        bidragsevneberegning.beregn(grunnlagBeregning).getResultatEvneBelop());

  }
}

File: src/test/kotlin/no/nav/bidrag/beregn/bidragsevne/BidragsevneBeregningTest.kt
package no.nav.bidrag.beregn.bidragsevne

import no.nav.bidrag.beregn.TestUtil
import no.nav.bidrag.beregn.bidragsevne.beregning.BidragsevneBeregningImpl
import no.nav.bidrag.beregn.bidragsevne.bo.BarnIHusstand
import no.nav.bidrag.beregn.bidragsevne.bo.Bostatus
import no.nav.bidrag.beregn.bidragsevne.bo.GrunnlagBeregning
import no.nav.bidrag.beregn.bidragsevne.bo.Inntekt
import no.nav.bidrag.beregn.bidragsevne.bo.Saerfradrag
import no.nav.bidrag.beregn.bidragsevne.bo.Skatteklasse
import no.nav.bidrag.beregn.felles.bo.Periode
import no.nav.bidrag.beregn.felles.bo.Sjablon
import no.nav.bidrag.beregn.felles.bo.SjablonInnhold
import no.nav.bidrag.beregn.felles.bo.SjablonPeriode
import no.nav.bidrag.beregn.felles.util.SjablonUtil
import no.nav.bidrag.domain.enums.BostatusKode
import no.nav.bidrag.domain.enums.InntektType
import no.nav.bidrag.domain.enums.SaerfradragKode
import no.nav.bidrag.domain.enums.sjablon.SjablonInnholdNavn
import no.nav.bidrag.domain.enums.sjablon.SjablonTallNavn
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import java.math.BigDecimal
import java.time.LocalDate

internal class BidragsevneBeregningTest {

    private val sjablonListe = TestUtil.byggSjabloner()
    private val sjablonPeriodeListe = TestUtil.byggSjablonPeriodeListe()
    private val bidragsevneberegning = BidragsevneBeregningImpl()

    @ParameterizedTest
    @CsvSource(
        // Test på beregning med ulike inntekter
        "1000000, 1, ALENE, 1.0, INGEN, 31859",
        "520000, 1, ALENE, 1.0, INGEN, 8322",
        "666000, 1, ALENE, 3.0, INGEN, 8424",
        "480000, 1, ALENE, 0.0, INGEN, 9976",
        // Test på at beregnet bidragsevne blir satt til 0 når evne er negativ
        "100000, 1, MED_ANDRE, 3.0, HELT, 0",
        // Test av halvt særfradrag
        "666000, 1, ALENE, 3.0, HALVT, 8965",
        // Test av bostatus MED_FLERE
        "666000, 1, MED_ANDRE, 3.0, HALVT, 14253"
    )
    fun testBidragsevneBeregningStandardSjabloner(
        inntektBelop: BigDecimal,
        skatteklasse: Int,
        bostatusKode: BostatusKode,
        antallBarn: Double,
        saerfradragKode: SaerfradragKode,
        expectedResult: Int
    ) {
        val inntekter = listOf(Inntekt(TestUtil.INNTEKT_REFERANSE, InntektType.LONN_SKE, inntektBelop))
        val grunnlagBeregning = GrunnlagBeregning(
            inntektListe = inntekter,
            skatteklasse = Skatteklasse(TestUtil.SKATTEKLASSE_REFERANSE, skatteklasse),
            bostatus = Bostatus(TestUtil.BOSTATUS_REFERANSE, bostatusKode),
            barnIHusstand = BarnIHusstand(TestUtil.BARN_I_HUSSTAND_REFERANSE, antallBarn),
            saerfradrag = Saerfradrag(TestUtil.SAERFRADRAG_REFERANSE, saerfradragKode),
            sjablonListe = sjablonPeriodeListe
        )
        val result = bidragsevneberegning.beregn(grunnlagBeregning).belop
        assertEquals(expectedResult.toBigDecimal(), result)
    }

    @Test
    fun testBidragsevneBeregningSpesifikkeSjabloner() {
        val inntekter = mutableListOf<Inntekt>()

        // Test at fordel skatteklasse 2 ikke legges til beregnet evne når skatteklasse = 1
        inntekter.add(
            Inntekt(
                referanse = TestUtil.INNTEKT_REFERANSE,
                inntektType = InntektType.LONN_SKE,
                inntektBelop = BigDecimal.valueOf(666000)
            )
        )
        sjablonPeriodeListe[0] = SjablonPeriode(
            sjablonPeriode = Periode(datoFom = LocalDate.parse("2017-01-01"), datoTil = LocalDate.parse("9999-12-31")),
            sjablon = Sjablon(
                navn = SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELOP.navn,
                nokkelListe = emptyList(),
                innholdListe = listOf(SjablonInnhold(navn = SjablonInnholdNavn.SJABLON_VERDI.navn, verdi = BigDecimal.valueOf(12000)))
            )
        )
        var grunnlagBeregning = GrunnlagBeregning(
            inntektListe = inntekter,
            skatteklasse = Skatteklasse(referanse = TestUtil.SKATTEKLASSE_REFERANSE, skatteklasse = 1),
            bostatus = Bostatus(referanse = TestUtil.BOSTATUS_REFERANSE, kode = BostatusKode.ALENE),
            barnIHusstand = BarnIHusstand(referanse = TestUtil.BARN_I_HUSSTAND_REFERANSE, antallBarn = 3.0),
            saerfradrag = Saerfradrag(referanse = TestUtil.SAERFRADRAG_REFERANSE, kode = SaerfradragKode.INGEN),
            sjablonListe = sjablonPeriodeListe
        )
        assertEquals(BigDecimal.valueOf(8424), bidragsevneberegning.beregn(grunnlagBeregning).belop)

        // Test at fordel skatteklasse 2 legges til beregnet evne når skatteklasse = 2
        inntekter[0] = Inntekt(
            referanse = TestUtil.INNTEKT_REFERANSE,
            inntektType = InntektType.LONN_SKE,
            inntektBelop = BigDecimal.valueOf(666000)
        )
        sjablonPeriodeListe[0] = SjablonPeriode(
            sjablonPeriode = Periode(datoFom = LocalDate.parse("2017-01-01"), datoTil = LocalDate.parse("9999-12-31")),
            sjablon = Sjablon(
                navn = SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELOP.navn,
                nokkelListe = emptyList(),
                innholdListe = listOf(SjablonInnhold(navn = SjablonInnholdNavn.SJABLON_VERDI.navn, verdi = BigDecimal.valueOf(12000)))
            )
        )
        grunnlagBeregning = GrunnlagBeregning(
            inntektListe = inntekter,
            skatteklasse = Skatteklasse(referanse = TestUtil.SKATTEKLASSE_REFERANSE, skatteklasse = 2),
            bostatus = Bostatus(referanse = TestUtil.BOSTATUS_REFERANSE, kode = BostatusKode.ALENE),
            barnIHusstand = BarnIHusstand(referanse = TestUtil.BARN_I_HUSSTAND_REFERANSE, antallBarn = 3.0),
            saerfradrag = Saerfradrag(referanse = TestUtil.SAERFRADRAG_REFERANSE, kode = SaerfradragKode.INGEN),
            sjablonListe = sjablonPeriodeListe
        )
        assertEquals(BigDecimal.valueOf(9424), bidragsevneberegning.beregn(grunnlagBeregning).belop)

        // Test at personfradrag skatteklasse 2 brukes hvis skatteklasse 2 er angitt
        sjablonPeriodeListe[0] = SjablonPeriode(
            sjablonPeriode = Periode(datoFom = LocalDate.parse("2017-01-01"), datoTil = LocalDate.parse("9999-12-31")),
            sjablon = Sjablon(
                navn = SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELOP.navn,
                nokkelListe = emptyList(),
                innholdListe = listOf(SjablonInnhold(navn = SjablonInnholdNavn.SJABLON_VERDI.navn, verdi = BigDecimal.ZERO))
            )
        )
        sjablonPeriodeListe[1] = SjablonPeriode(
            sjablonPeriode = Periode(datoFom = LocalDate.parse("2017-01-01"), datoTil = LocalDate.parse("9999-12-31")),
            sjablon = Sjablon(
                navn = SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELOP.navn,
                nokkelListe = emptyList(),
                innholdListe = listOf(SjablonInnhold(navn = SjablonInnholdNavn.SJABLON_VERDI.navn, verdi = BigDecimal.valueOf(24000)))
            )
        )
        grunnlagBeregning = GrunnlagBeregning(
            inntektListe = inntekter,
            skatteklasse = Skatteklasse(referanse = TestUtil.SKATTEKLASSE_REFERANSE, skatteklasse = 2),
            bostatus = Bostatus(referanse = TestUtil.BOSTATUS_REFERANSE, kode = BostatusKode.ALENE),
            barnIHusstand = BarnIHusstand(referanse = TestUtil.BARN_I_HUSSTAND_REFERANSE, antallBarn = 3.0),
            saerfradrag = Saerfradrag(referanse = TestUtil.SAERFRADRAG_REFERANSE, kode = SaerfradragKode.INGEN),
            sjablonListe = sjablonPeriodeListe
        )
        assertEquals(BigDecimal.valueOf(7923), bidragsevneberegning.beregn(grunnlagBeregning).belop)
    }

    @Test
    fun beregnMinstefradrag() {
        val inntekter = mutableListOf<Inntekt>()

        inntekter.add(
            Inntekt(
                referanse = TestUtil.INNTEKT_REFERANSE,
                inntektType = InntektType.LONN_SKE,
                inntektBelop = BigDecimal.valueOf(200000)
            )
        )
        var grunnlagBeregning = GrunnlagBeregning(
            inntektListe = inntekter,
            skatteklasse = Skatteklasse(referanse = TestUtil.SKATTEKLASSE_REFERANSE, skatteklasse = 1),
            bostatus = Bostatus(referanse = TestUtil.BOSTATUS_REFERANSE, kode = BostatusKode.ALENE),
            barnIHusstand = BarnIHusstand(referanse = TestUtil.BARN_I_HUSSTAND_REFERANSE, antallBarn = 1.0),
            saerfradrag = Saerfradrag(referanse = TestUtil.SAERFRADRAG_REFERANSE, kode = SaerfradragKode.HELT),
            sjablonListe = sjablonPeriodeListe
        )

        assertThat(
            bidragsevneberegning.beregnMinstefradrag(
                grunnlag = grunnlagBeregning,
                minstefradragInntektSjablonBelop = SjablonUtil.hentSjablonverdi(
                    sjablonListe = sjablonListe,
                    sjablonTallNavn = SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP
                ),
                minstefradragInntektSjablonProsent = SjablonUtil.hentSjablonverdi(
                    sjablonListe = sjablonListe,
                    sjablonTallNavn = SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT
                )
            )
        )
            .isEqualTo(BigDecimal.valueOf(62000))

        inntekter[0] = Inntekt(
            referanse = TestUtil.INNTEKT_REFERANSE,
            inntektType = InntektType.LONN_SKE,
            inntektBelop = BigDecimal.valueOf(1000000)
        )
        grunnlagBeregning = GrunnlagBeregning(
            inntektListe = inntekter,
            skatteklasse = Skatteklasse(referanse = TestUtil.SKATTEKLASSE_REFERANSE, skatteklasse = 1),
            bostatus = Bostatus(referanse = TestUtil.BOSTATUS_REFERANSE, kode = BostatusKode.ALENE),
            barnIHusstand = BarnIHusstand(referanse = TestUtil.BARN_I_HUSSTAND_REFERANSE, antallBarn = 1.0),
            saerfradrag = Saerfradrag(referanse = TestUtil.SAERFRADRAG_REFERANSE, kode = SaerfradragKode.HELT),
            sjablonListe = sjablonPeriodeListe
        )

        assertThat(
            bidragsevneberegning.beregnMinstefradrag(
                grunnlag = grunnlagBeregning,
                minstefradragInntektSjablonBelop = SjablonUtil.hentSjablonverdi(
                    sjablonListe = sjablonListe,
                    sjablonTallNavn = SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP
                ),
                minstefradragInntektSjablonProsent = SjablonUtil.hentSjablonverdi(
                    sjablonListe = sjablonListe,
                    sjablonTallNavn = SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT
                )
            )
        )
            .isEqualTo(BigDecimal.valueOf(87450))
    }

    @Test
    fun beregnSkattetrinnBelop() {
        val inntekter = mutableListOf<Inntekt>()

        inntekter.add(
            Inntekt(
                referanse = TestUtil.INNTEKT_REFERANSE,
                inntektType = InntektType.LONN_SKE,
                inntektBelop = BigDecimal.valueOf(666000)
            )
        )
        var grunnlagBeregning = GrunnlagBeregning(
            inntektListe = inntekter,
            skatteklasse = Skatteklasse(referanse = TestUtil.SKATTEKLASSE_REFERANSE, skatteklasse = 1),
            bostatus = Bostatus(referanse = TestUtil.BOSTATUS_REFERANSE, kode = BostatusKode.ALENE),
            barnIHusstand = BarnIHusstand(referanse = TestUtil.BARN_I_HUSSTAND_REFERANSE, antallBarn = 1.0),
            saerfradrag = Saerfradrag(referanse = TestUtil.SAERFRADRAG_REFERANSE, kode = SaerfradragKode.HELT),
            sjablonListe = sjablonPeriodeListe
        )
        assertEquals(BigDecimal.valueOf((1400 + 16181 + 3465 + 0).toLong()), bidragsevneberegning.beregnSkattetrinnBelop(grunnlagBeregning))

        inntekter[0] = Inntekt(
            referanse = TestUtil.INNTEKT_REFERANSE,
            inntektType = InntektType.LONN_SKE,
            inntektBelop = BigDecimal.valueOf(174600)
        )
        grunnlagBeregning = GrunnlagBeregning(
            inntektListe = inntekter,
            skatteklasse = Skatteklasse(referanse = TestUtil.SKATTEKLASSE_REFERANSE, skatteklasse = 1),
            bostatus = Bostatus(referanse = TestUtil.BOSTATUS_REFERANSE, kode = BostatusKode.ALENE),
            barnIHusstand = BarnIHusstand(referanse = TestUtil.BARN_I_HUSSTAND_REFERANSE, antallBarn = 1.0),
            saerfradrag = Saerfradrag(referanse = TestUtil.SAERFRADRAG_REFERANSE, kode = SaerfradragKode.HELT),
            sjablonListe = sjablonPeriodeListe
        )
        assertEquals(BigDecimal.ZERO, bidragsevneberegning.beregnSkattetrinnBelop(grunnlagBeregning))

        inntekter[0] = Inntekt(
            referanse = TestUtil.INNTEKT_REFERANSE,
            inntektType = InntektType.LONN_SKE,
            inntektBelop = BigDecimal.valueOf(250000)
        )
        grunnlagBeregning = GrunnlagBeregning(
            inntektListe = inntekter,
            skatteklasse = Skatteklasse(referanse = TestUtil.SKATTEKLASSE_REFERANSE, skatteklasse = 1),
            bostatus = Bostatus(referanse = TestUtil.BOSTATUS_REFERANSE, kode = BostatusKode.ALENE),
            barnIHusstand = BarnIHusstand(referanse = TestUtil.BARN_I_HUSSTAND_REFERANSE, antallBarn = 1.0),
            saerfradrag = Saerfradrag(referanse = TestUtil.SAERFRADRAG_REFERANSE, kode = SaerfradragKode.HELT),
            sjablonListe = sjablonPeriodeListe
        )
        assertEquals(BigDecimal.valueOf(1315), bidragsevneberegning.beregnSkattetrinnBelop(grunnlagBeregning))
    }
}
